### PR TITLE
Rewrite README

### PR DESCRIPTION
**Pull request by ewindisch**
Signed-off-by: Eric Windisch eric@windisch.us

### CODE REVIEW

**Comment by ewindisch** [comment]
cc @pselle

File: README.md [hunk 1]
```diff
@@ -5,25 +5,39 @@ IOpipe
 Apache 2.0 licensed.
 
 IOpipe is a toolkit for building and orchestrating event-driven and
-serverless applications. These apps may run *locally* or in the cloud
-via AWS Lambda, Google Cloud Functions, or Azure Functions.
+serverless applications. These apps may run anywhere, either *locally* or,
```
Comment (pselle) [inline README.md]: why is locally in italics?

File: README.md [hunk 2]
```diff
@@ -5,25 +5,39 @@ IOpipe
...
+via execution drivers, in the cloud.
```
Comment (pselle) [inline README.md:9]: I see how you wanted to remove specific names, but I immediately thought "wtf is an execution driver."  FaaS? On-demand compute services? (EDIT) now I see the "exist for" below. 

Maybe "or use our execution drivers to run your code in the cloud"

File: README.md [hunk 3]
```diff
@@ -5,25 +5,39 @@ IOpipe
...
 input and communicate over the networking, operating in a fashion to
 Unix pipes. A kernel may receive input or send output to/from
 web service requests, functions, or local applications.
 
+# CLI
 
----------------------------------------
-Usage
----------------------------------------
+A CLI exists to create and export npm modules, share code,
```
Comment (pselle) [inline README.md]: Use our CLI (link to it) to create and export ...

File: README.md [hunk 4]
```diff
@@ -5,25 +5,39 @@ IOpipe
...
+and provide runtime of magnetic kernels. This project is written in Go.
```
Comment (pselle) [inline README.md]: Doesn't matter that it's written in Go. Does the user have to care? If they can download it and install it right from the commandline, give them the oneliner to do so right here

File: README.md [hunk 5]
```diff
@@ -5,25 +5,39 @@ IOpipe
...
+
+Find, download, and/or contribute to this tool in the [IOpipe-Golang repo](https://github.com/iopipe/iopipe-golang).
+
+# SDK
 
 ### NodeJS SDK:
```
Comment (pselle) [inline README.md:37]: If there's only 1 SDK, do we need to specify/point out that it's Node?

File: README.md [hunk 6]
```diff
@@ -36,53 +50,60 @@ The callback variable received by a function is *also* an AWS Lambda-compatible
...
-                               iopipe.fetch,
-                               (event, callback) => {
-                                  callback(crypto
-                                           .createHash('sha256')
-                                           .update(event)
-                                           .digest('hex'))
-                               })
+exports.handle = iopipe.define(
+  (event, context) => {
+    // The succeed function could be omitted by simply calling context()
```
Comment (pselle) [inline README.md]: Does this comment matter or help the user?

File: README.md [hunk 7]
```diff
@@ -36,53 +50,60 @@ The callback variable received by a function is *also* an AWS Lambda-compatible
...
+  }
+)
+```
+
+#### Indicating failures
+
+Use the context.fail function to pass errors.
+
+```javascript
+/* Create a Lambda function which increments event.key by 1. */
```
Comment (pselle) [inline README.md]: This is not the right comment for this block

File: README.md [hunk 8]
```diff
@@ -36,53 +50,60 @@ The callback variable received by a function is *also* an AWS Lambda-compatible
...
+  }
+)
+```
+
+#### HTTP endpoints as "functions"
+
+```javascript
+/* Lambda function which fetches data from a URL & does a POST to another. */
+exports.handle = iopipe.define("http://localhost/get-data",
+														   "http://localhost/post-data")
```
Comment (pselle) [inline README.md:98]: Not a super fan of this spacing, it looks a little weird

File: README.md [hunk 9]
```diff
@@ -36,53 +50,60 @@ The callback variable received by a function is *also* an AWS Lambda-compatible
...
+      context.fail(err)
+    }
+  }
+)
+```
+
+#### HTTP endpoints as "functions"
+
+```javascript
+/* Lambda function which fetches data from a URL & does a POST to another. */
```
Comment (pselle) [inline README.md]: _then_ does a POST to another (passing the data from the first request?)

File: README.md [hunk 10]
```diff
@@ -36,53 +50,60 @@ The callback variable received by a function is *also* an AWS Lambda-compatible
...
+exports.handle = iopipe.define("http://localhost/get-data",
+														   "http://localhost/post-data")
+```
+
+```javascript
+// Users may chain HTTP requests & other functions.
```
Comment (pselle) [inline README.md]: Users may => You can

File: README.md [hunk 11]
```diff
@@ -36,53 +50,60 @@ The callback variable received by a function is *also* an AWS Lambda-compatible
...
+exports.handle = iopipe.define("http://localhost/get-data",
+                               (data, callback) => {
+                                 console.log("Fetched data: " + data)
+                               }
```
Comment (pselle) [inline README.md]: Spacing is weird here

File: README.md [hunk 12]
```diff
@@ -36,53 +50,60 @@ The callback variable received by a function is *also* an AWS Lambda-compatible
...
 
-### AWS Lambda Client
+### Scatter & Gather
+
+IOpipe also acts as a client to serverless infrastructure allowing
+the use of scatter & gather patterns such as map-reduce.
 
-IOpipe also acts as an AWS Lambda Client where a Lambda function may
+Below we initialize an AWS Lambda Client where a Lambda function may
 be specified by its URN and included in the execution chain:
```
Comment (pselle) [inline README.md]: Should we link URN to the definition of what that is?

File: README.md [hunk 13]
```diff
@@ -97,8 +118,8 @@ var iopipe_aws = require("iopipe")(
 )
 var crypto = require("crypto")
 
-export.handler = iopipe_aws.define("urn:somefunction",
-                                   "urn:anotherfunction",
+export.handler = iopipe_aws.define("urn:someLambdaFunction",
+                                   "urn:anotherLambdaFunction",
                                    iopipe.property("property-of-result"),
                                    iopipe.fetch, # fetch that as a URL
```
Comment (pselle) [inline README.md]: `#` is not a JS comment, change to //

File: README.md [hunk 14]
```diff
@@ -114,42 +135,9 @@ export.handler = iopipe_aws.define("urn:somefunction",
...
-
-Find this tool in the [IOpipe-Golang repo](https://github.com/iopipe/iopipe-golang).
-
----------------------------------------
+---------
 Security
----------------------------------------
+---------
 
 Kernels are executed in individual virtual machines
```
Comment (pselle) [inline README.md]: This is now updated to refer less to kernels ... is it still clear to a reader what they are?

File: README.md [hunk 15]
```diff
@@ -5,85 +5,122 @@ IOpipe
...
+
+ - context.done(err, data)
+ - context.succeed(data)
+ - context.fail(err)
+
+
+Example of using context.fail to pass errors:
+
+```javascript
+/* Create a Lambda function which increments event.key by 1. */
```
Comment (pselle) [inline README.md]: This comment is still incorrect for this block

File: README.md [hunk 16]
```diff
@@ -5,85 +5,122 @@ IOpipe
...
 ```
 
-### AWS Lambda Client
+#### Context argument
+
+The context argument operates as both a callback and
+an object with several methods, similar to the same
+argument passed to AWS Lambda functions.
+
+Developers may call context() directly, with its argument
```
Comment (pselle) [inline README.md]: call `context()` directly

File: README.md [hunk 17]
```diff
@@ -114,42 +151,14 @@ export.handler = iopipe_aws.define("urn:somefunction",
...
-Requests and responses are translated using kernels, and
-may pipe to other kernels, or to/from web service endpoints.
-
-Kernels simply receive request or response data and output
-translated request or response data.
+### Go SDK:
 
-Example:
+Bundled with the [IOpipe CLI](https://github.com/iopipe/iopipe-golang) is
+a Go SDK, still in early-development.
```
Comment (pselle) [inline README.md]: early-development => early development

**Pull request merged**